packages/wdio-browserstack-service/src/testOps/listener.ts

```diff
@@ -0,0 +1,227 @@
+import UsageStats from './usageStats.js'
+import type FeatureStats from './featureStats.js'
+import RequestQueueHandler from '../request-handler.js'
+import type { CBTData, LogData, ScreenshotLog, TestData, UploadType } from '../types.js'
+import { batchAndPostEvents, sleep } from '../util.js'
+import {
+    DATA_BATCH_ENDPOINT,
+    DEFAULT_WAIT_INTERVAL_FOR_PENDING_UPLOADS,
+    DEFAULT_WAIT_TIMEOUT_FOR_PENDING_UPLOADS,
+    LOG_KIND_USAGE_MAP
+} from '../constants.js'
+import { sendScreenshots } from './requestUtils.js'
+import { BStackLogger } from '../bstackLogger.js'
+
+class Listener {
+    private static instance: Listener
+    private readonly usageStats: UsageStats = UsageStats.getInstance()
+    private readonly testStartedStats: FeatureStats = this.usageStats.testStartedStats
+    private readonly testFinishedStats: FeatureStats = this.usageStats.testFinishedStats
+    private readonly hookStartedStats: FeatureStats = this.usageStats.hookStartedStats
+    private readonly hookFinishedStats: FeatureStats = this.usageStats.hookFinishedStats
+    private readonly cbtSessionStats: FeatureStats = this.usageStats.cbtSessionStats
+    private readonly logEvents: FeatureStats = this.usageStats.logStats
+    private requestBatcher?: RequestQueueHandler
+    private pendingUploads = 0
+
+    private constructor() {
+    }
+
+    public static getInstance(): Listener {
+        if (!Listener.instance) {
+            Listener.instance = new Listener()
+        }
+        return Listener.instance
+    }
+
+    public async onWorkerEnd() {
+        try {
+            await this.uploadPending()
+            await this.teardown()
+        } catch (e) {
+            BStackLogger.debug('Exception in onWorkerEnd: ' + e)
+        }
+    }
+
+    async uploadPending(waitTimeout = DEFAULT_WAIT_TIMEOUT_FOR_PENDING_UPLOADS, waitInterval = DEFAULT_WAIT_INTERVAL_FOR_PENDING_UPLOADS): Promise<unknown> {
+        if ((this.pendingUploads <= 0) || waitTimeout <= 0) {
+            return
+        }
+
+        await sleep(waitInterval)
+        return this.uploadPending(waitTimeout - waitInterval)
+    }
+
+    async teardown() {
+        BStackLogger.debug('teardown started')
+        RequestQueueHandler.tearDownInvoked = true
+        await this.requestBatcher?.shutdown()
+        BStackLogger.debug('teardown ended')
+    }
+
+    public hookStarted(hookData: TestData): void {
+        try {
+            this.hookStartedStats.triggered()
+            this.sendBatchEvents(this.getEventForHook('HookRunStarted', hookData))
+        } catch (e) {
+            this.hookStartedStats.failed()
+            throw e
+        }
+    }
+
+    public hookFinished(hookData: TestData): void {
+        try {
+            this.hookFinishedStats.triggered(hookData.result)
+            this.sendBatchEvents(this.getEventForHook('HookRunFinished', hookData))
```

Comment:
same

Reply:
The `sendBatchEvents` method calls `add` method of RequestQueueHandler which can throw exception
- currently when build start not yet completed. 

This try catch is added to mainly get the proper instrumentation on how many events, we were able to send properly and how many failed. That's why this try catch is needed to catch the instrumentation properly, we are re-throwing the exception from here after catching.